File: legacy/v2-python/note_tools.py
import re
from datetime import datetime


def parse_date(input_str: str) -> datetime:
    """
    Parses a date from the given input string.
    
    The input can be in various formats:
    - "[[yyyy-MM-dd-DayOfWeek|dd.MM.yyyy]]"
    - "[[yyyy-MM-dd-DayOfWeek]]"
    - "dd.MM.yyyy"
    
    Args:
        input_str: The input string containing the date to parse.
        
    Returns:
        The parsed datetime object.
        
    Raises:
        ValueError: If the input string is not in a valid date format.
    """
    # Remove leading '#' characters and whitespace
    input_str = input_str.lstrip('#').strip()
    
    if input_str.startswith("[[") and input_str.endswith("]]"):
        content = input_str[2:-2]  # Remove "[[" and "]]"
        
        if "|" in content:
            # Split the content on '|'
            parts = content.split('|')
            date_str1 = parts[0]  # "yyyy-MM-dd-DayOfWeek"
            date_str2 = parts[1]  # "dd.MM.yyyy"
            
            # Extract "yyyy-MM-dd" from date_str1
            date_part1_tokens = date_str1.split('-')
            if len(date_part1_tokens) >= 3:
                date_part1 = "-".join(date_part1_tokens[0:3])
                try:
                    return datetime.strptime(date_part1, "%Y-%m-%d")
                except ValueError:
                    pass
            
            # Try parsing date_str2
            try:
                return datetime.strptime(date_str2, "%d.%m.%Y")
            except ValueError:
                raise ValueError("Invalid date format in input with '|'")
        else:
            # Content is "yyyy-MM-dd-DayOfWeek"
            date_part_tokens = content.split('-')
            if len(date_part_tokens) >= 3:
                date_part = "-".join(date_part_tokens[0:3])
                try:
                    return datetime.strptime(date_part, "%Y-%m-%d")
                except ValueError:
                    raise ValueError("Invalid date format in input without '|'")
            raise ValueError("Invalid date format in input without '|'")
    else:
        # Try parsing as "dd.MM.yyyy"
        try:
            return datetime.strptime(input_str, "%d.%m.%Y")
        except ValueError:
            raise ValueError("Invalid date format in simple input")

File: legacy/v1-csharp/EnglishFlashcardGenerator/NoteTools.cs
namespace EnglishFlashcardGenerator;

public static class NoteTools
{
    /// <summary>
    /// Parses a date from the given input string.
    /// </summary>
    /// <param name="input">The input string containing the date to parse. The input can be in various formats:
    /// <list type="bullet">
    /// <item><description>"[[yyyy-MM-dd-DayOfWeek|dd.MM.yyyy]]"</description></item>
    /// <item><description>"[[yyyy-MM-dd-DayOfWeek]]"</description></item>
    /// <item><description>"dd.MM.yyyy"</description></item>
    /// </list>
    /// </param>
    /// <returns>The parsed <see cref="DateTime"/> object.</returns>
    /// <exception cref="FormatException">Thrown when the input string is not in a valid date format.</exception>
    public static DateTime ParseDate(string input)
    {
        // Remove leading '#' characters and whitespace
        input = input.TrimStart('#', ' ').Trim();

        if (input.StartsWith("[[") && input.EndsWith("]]"))
        {
            var content = input.Substring(2, input.Length - 4); // Remove "[[" and "]]"

            if (content.Contains("|"))
            {
                // Split the content on '|'
                var parts = content.Split('|');
                var dateStr1 = parts[0]; // "yyyy-MM-dd-DayOfWeek"
                var dateStr2 = parts[1]; // "dd.MM.yyyy"

                // Extract "yyyy-MM-dd" from dateStr1
                var datePart1Tokens = dateStr1.Split('-');
                if (datePart1Tokens.Length >= 3)
                {
                    var datePart1 = string.Join("-", datePart1Tokens[0], datePart1Tokens[1], datePart1Tokens[2]);
                    if (DateTime.TryParseExact(datePart1, "yyyy-MM-dd", null, System.Globalization.DateTimeStyles.None, out DateTime date))
                    {
                        return date;
                    }
                }

                // Try parsing dateStr2
                if (DateTime.TryParseExact(dateStr2, "dd.MM.yyyy", null, System.Globalization.DateTimeStyles.None, out DateTime date2))
                {
                    return date2;
                }

                throw new FormatException("Invalid date format in input with '|'");
            }
            else
            {
                // Content is "yyyy-MM-dd-DayOfWeek"
                var datePartTokens = content.Split('-');
                if (datePartTokens.Length >= 3)
                {
                    var datePart = string.Join("-", datePartTokens[0], datePartTokens[1], datePartTokens[2]);
                    if (DateTime.TryParseExact(datePart, "yyyy-MM-dd", null, System.Globalization.DateTimeStyles.None, out DateTime date))
                    {
                        return date;
                    }
                }
                throw new FormatException("Invalid date format in input without '|'");
            }
        }
        else
        {
            // Try parsing as "dd.MM.yyyy"
            if (DateTime.TryParseExact(input, "dd.MM.yyyy", null, System.Globalization.DateTimeStyles.None, out DateTime date))
            {
                return date;
            }
            else
            {
                throw new FormatException("Invalid date format in simple input");
            }
        }
    }
    
}